From 0fc2d593b7c68f996989bd16daef159882a7041e Mon Sep 17 00:00:00 2001
From: Rob Crittenden <rcritten@redhat.com>
Date: Wed, 3 Sep 2025 14:37:45 -0400
Subject: [PATCH 3/6] Add initial ML-DSA support with NSS 3.112.0-4

NSS 3.112 has added support for Post-Quantum ML-DSA keys.
Allow them to be generated by certmonger.

The minimum key size is not fully implemented because key size
has a different meaning in this context. It is plumbed in, such that
one can set a numeric size, but it isn't used and is effectively
meaningless.

This includes what may be a temporary change to identify the key
type when reading a private key. NSS has an internal function
for this that may be exposed in the future. This works for now.
---
 src/keygen-n.c                       | 26 +++++++++++-
 src/keyiread-n.c                     | 59 +++++++++++++++++++++++++++-
 tests/002-keygen-ml-dsa/expected.out | 13 ++++++
 tests/002-keygen-ml-dsa/prequal.sh   |  5 +++
 tests/002-keygen-ml-dsa/run.sh       | 36 +++++++++++++++++
 5 files changed, 136 insertions(+), 3 deletions(-)
 create mode 100644 tests/002-keygen-ml-dsa/prequal.sh

diff --git a/src/keygen-n.c b/src/keygen-n.c
index 3c3b6cd0..45b1426b 100644
--- a/src/keygen-n.c
+++ b/src/keygen-n.c
@@ -139,6 +139,9 @@ cm_keygen_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 #ifdef CM_ENABLE_EC
 	SECOidData *ecurve;
 	SECItem ec_params;
+#endif
+#ifdef CM_ENABLE_ML_DSA
+	CK_ULONG ml_params;
 #endif
 	SECKEYPrivateKey *privkey, *delkey, *ckey;
 	SECKEYPrivateKeyList *privkeys;
@@ -273,8 +276,15 @@ cm_keygen_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 #endif
 #ifdef CM_ENABLE_ML_DSA:
 	case cm_key_ml_dsa_44:
+		cm_requested_key_size = ML_DSA_44_PRIVATEKEY_LEN;
+		mech = CKM_ML_DSA_KEY_PAIR_GEN;
+		pmech = CKM_ML_DSA_KEY_PAIR_GEN;
 	case cm_key_ml_dsa_65:
+		cm_requested_key_size = ML_DSA_65_PRIVATEKEY_LEN;
+		mech = CKM_ML_DSA_KEY_PAIR_GEN;
+		pmech = CKM_ML_DSA_KEY_PAIR_GEN;
 	case cm_key_ml_dsa_87:
+		cm_requested_key_size = ML_DSA_87_PRIVATEKEY_LEN;
 		mech = CKM_ML_DSA_KEY_PAIR_GEN;
 		pmech = CKM_ML_DSA_KEY_PAIR_GEN;
 		break;
@@ -330,6 +340,8 @@ next_slot:
 		cm_log(1, "Error locating token for key generation.\n");
 		_exit(CM_SUB_STATUS_ERROR_NO_TOKEN);
 	}
+	/* FIXME */
+	cm_log(1, "Selected token %s\n", PK11_GetTokenName(slot));
 	/* Be ready to count our uses of a PIN. */
 	memset(&cb_data, 0, sizeof(cb_data));
 	cb_data.entry = entry;
@@ -582,9 +594,19 @@ retry_gen:
 #endif
 #ifdef CM_ENABLE_ML_DSA:
 	case cm_key_ml_dsa_44:
+		memset(&ml_params, 0, sizeof(ml_params));
+		ml_params = CKP_ML_DSA_44;
+		params = &ml_params;
+		break;
 	case cm_key_ml_dsa_65:
+		memset(&ml_params, 0, sizeof(ml_params));
+		ml_params = CKP_ML_DSA_65;
+		params = &ml_params;
+		break;
 	case cm_key_ml_dsa_87:
-		/* FIXME */
+		memset(&ml_params, 0, sizeof(ml_params));
+		ml_params = CKP_ML_DSA_87;
+		params = &ml_params;
 		break;
 #endif
 	default:
@@ -592,7 +614,7 @@ retry_gen:
 		break;
 	}
 	/* Generate the key pair. */
-	cm_log(1, "Generating key pair.\n");
+	cm_log(1, "Generating key pair of size %d.\n", cm_key_size);
 	pubkey = NULL;
 	privkey = PK11_GenerateKeyPair(slot, mech, params, &pubkey,
 				       PR_TRUE, PR_TRUE, NULL);
diff --git a/src/keyiread-n.c b/src/keyiread-n.c
index ed87b095..7f9cdf09 100644
--- a/src/keyiread-n.c
+++ b/src/keyiread-n.c
@@ -52,6 +52,60 @@
 #define PRIVKEY_LIST_EMPTY(l) PRIVKEY_LIST_END(PRIVKEY_LIST_HEAD(l), l)
 #endif
 
+
+/* FIXME: replace when NSS exposes the capability to get the alg */
+/* look up the an oid by string.
+ *
+ * The following code was ripped from NSS.
+ * */
+#include <cryptohi.h>
+
+/* If SECKEY_GetParameterSet() is exported then this can be replaced */
+const char *
+cm_GetSignatureAlgorithmFromPrivateKey(SECKEYPrivateKey *privKey)
+{
+    SECItem item;
+    CK_ULONG paramSet;
+    SECStatus rv;
+
+    /* don't modify hashAlg if we aren't a DSA key */
+    if (privKey->keyType != mldsaKey) {
+	/* should not get here */
+        return "Invalid ML-DSA key";
+    }
+
+    rv = PK11_ReadRawAttribute(PK11_TypePrivKey, privKey,
+                                CKA_PARAMETER_SET, &item);
+
+    if (rv != SECSuccess) {
+        return "UNKNOWN";
+    }
+    if (item.len != sizeof (paramSet)) {
+        PORT_Free(item.data);
+        PORT_SetError(SEC_ERROR_INVALID_KEY);
+        return "UNKNOWN";
+    }
+    paramSet = *(CK_ULONG *)item.data;
+    PORT_Free(item.data);
+    switch  (paramSet) {
+        case CKP_ML_DSA_44:
+            return "ML-DSA-44";
+            break;
+        case CKP_ML_DSA_65:
+            return "ML-DSA-65";
+            break;
+        case CKP_ML_DSA_87:
+            return "ML-DSA-87";
+            break;
+        default:
+            return "UNKNOWN";
+            break;
+    }
+    return "UNKNOWN";
+}
+
+/* end FIXME */
+
 struct cm_keyiread_state {
 	struct cm_keyiread_state_pvt pvt;
 	struct cm_store_entry *entry;
@@ -522,7 +576,10 @@ cm_keyiread_n_main(int fd, struct cm_store_ca *ca, struct cm_store_entry *entry,
 			cm_log(3, "Key is an EC key.\n");
 			alg = "EC";
 			break;
-		/* FIXME: case mlKey? */
+		case mldsaKey:
+			cm_log(3, "Key is an ML-DSA key.\n");
+			alg = cm_GetSignatureAlgorithmFromPrivateKey(keys->privkey);
+			break;
 		case nullKey:
 		default:
 			cm_log(3, "Key is of an unknown type.\n");
diff --git a/tests/002-keygen-ml-dsa/expected.out b/tests/002-keygen-ml-dsa/expected.out
index b3039590..88da9f1e 100644
--- a/tests/002-keygen-ml-dsa/expected.out
+++ b/tests/002-keygen-ml-dsa/expected.out
@@ -1,3 +1,16 @@
+[nss:44]
+OK.
+OK (ML-DSA-44:10496).
+[nss:65]
+OK.
+OK (ML-DSA-65:15616).
+[nss:87]
+OK.
+OK (ML-DSA-87:20736).
+[nss:rosubdir]
+Failed to save NSS:${tmpdir}/rosubdir: need fs permissions.
+[nss:rwsubdir]
+Failed to save NSS:${tmpdir}/rwsubdir: need fs permissions.
 [openssl:44]
 OK.
 OK (ML-DSA-44:10496).
diff --git a/tests/002-keygen-ml-dsa/prequal.sh b/tests/002-keygen-ml-dsa/prequal.sh
new file mode 100644
index 00000000..d146a650
--- /dev/null
+++ b/tests/002-keygen-ml-dsa/prequal.sh
@@ -0,0 +1,5 @@
+#!/bin/sh
+if test `id -u` -eq 0 ; then
+	echo "This test won't work right if run as root."
+	exit 1
+fi
diff --git a/tests/002-keygen-ml-dsa/run.sh b/tests/002-keygen-ml-dsa/run.sh
index 55360f7d..6adc541c 100755
--- a/tests/002-keygen-ml-dsa/run.sh
+++ b/tests/002-keygen-ml-dsa/run.sh
@@ -5,6 +5,42 @@ cd "$tmpdir"
 source "$srcdir"/functions
 initnssdb "$tmpdir"
 
+for size in 44 65 87 ; do
+	echo "[nss:$size]"
+	# Generate a key.
+	cat > entry.$size <<- EOF
+	key_storage_type=NSSDB
+	key_storage_location=$tmpdir
+	key_nickname=keyi$size
+	key_gen_size=$size
+	key_gen_type=ML-DSA-$size
+	EOF
+	$toolsdir/keygen entry.$size
+	# Read the type and size.
+	sed -i 's,^key_gen_size.*,,g' entry.$size
+	$toolsdir/keyiread entry.$size
+done
+
+echo "[nss:rosubdir]"
+cat > entry.$size <<- EOF
+key_storage_type=NSSDB
+key_storage_location=$tmpdir/rosubdir
+key_nickname=keyi$size
+key_gen_size=$size
+key_gen_type=ML-DSA-$size
+EOF
+$toolsdir/keygen entry.$size || true
+
+echo "[nss:rwsubdir]"
+cat > entry.$size <<- EOF
+key_storage_type=NSSDB
+key_storage_location=$tmpdir/rwsubdir
+key_nickname=keyi$size
+key_gen_size=$size
+key_gen_type=ML-DSA-$size
+EOF
+$toolsdir/keygen entry.$size || true
+
 for size in 44 65 87 ; do
 	echo "[openssl:$size]"
 	# Generate a key.
-- 
2.52.0

